controller/cmd/service-mirror/probe_worker.go

```diff
@@ -0,0 +1,124 @@
+package servicemirror
+
+import (
+	"fmt"
+	"math/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	logging "github.com/sirupsen/logrus"
+)
+
+// ProbeWorker is responsible for monitoring gateways using a probe specification
+type ProbeWorker struct {
+	*sync.RWMutex
+	probe                 *GatewayProbeSpec
+	numAssociatedServices int32
+	stopCh                chan struct{}
+	metrics               *probeMetrics
+	log                   *logging.Entry
+}
+
+// NewProbeWorker creates a new probe worker associated with a particular gateway
+func NewProbeWorker(probe *GatewayProbeSpec, metrics *probeMetrics, probekey string) *ProbeWorker {
+	return &ProbeWorker{
+		RWMutex:               &sync.RWMutex{},
+		probe:                 probe,
+		numAssociatedServices: 0,
+		stopCh:                make(chan struct{}, 1),
+		metrics:               metrics,
+		log: logging.WithFields(logging.Fields{
+			"probe-key": probekey,
+		}),
+	}
+}
+
+// IncNumServices increments the number of services that are routed by the gateway
+func (pw *ProbeWorker) IncNumServices() {
+	pw.numAssociatedServices = pw.numAssociatedServices + 1
```

Comment:
If we need to worry about synchronizing probe spec updates, should also do that for incrementing and decrementing the services?

Reply:
@kleimkuhler  Not sure we need to take the lock here. `IncNumServices` and as a result `numAssociatedServices` are only called/written/read from within probe_manager.go which processes events one at a time from its chan. The reason we synchronize the probe spec updates is because the latter can be modified by the probe_manager and are also read by the go routine that calls `doProbe()`. Does that sound right ? WDYT?

Reply:
Yep makes sense.